### PR TITLE
Add update review action

### PR DESCRIPTION
**Pull request by gcsecsey**
Fixes https://github.com/Automattic/dotcom-forge/issues/3921

## Proposed Changes

 - add editing state to review item
 - add state to store updated review content
 - call update endpoint with the updated rating/content

## Known issues

The API currently serves the reviews with HTML formatting. We need to restrict the reviews to be plain-text in a follow-up.

## Testing Instructions

- add a review to WooCommerce Bookings by going to [http://calypso.localhost:3000/marketplace/test/[siteId]](http://calypso.localhost:3000/marketplace/test/%5BsiteId%5D)
- go to `/plugins/woocommerce-bookings`
- click read all reviews, then the `Edit my review` button
- check the styling of cancel / save buttons
- disable the `marketplace-reviews-notifications` feature flag by adding the `?flags=-marketplace-reviews-notifications` parameter to the URL
- reload the page and check that the notify checkbox is hidden
  CleanShot 2024-01-03 at 15 49 06@2x
- update the rating and the review text and click `Save my review`
- Check that the editing view is closed and the review is updated in the list

CleanShot 2023-12-29 at 12 56 26

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by gcsecsey** [comment]
For handover/posterity: I rebased trunk, and the task https://github.com/Automattic/wp-calypso/pull/85323 has replaced the reviews list with the cards, so currently the reviews list is not rendered.

As this change would need the reviews list to be displayed, the component can only be updated when the 23911c187a version is checked out. cc @WBerredo

**Comment by WBerredo** [comment]
Given the reviews list is needed, I'm bringing this task https://github.com/Automattic/dotcom-forge/issues/4843 upfront to unblock all tasks that need the list.

**Comment by cpapazoglou** [comment]
After https://github.com/Automattic/wp-calypso/pull/85828 is merged, this will need a rebase

File: client/my-sites/marketplace/components/reviews-list/index.tsx
Comment (cpapazoglou) [inline client/my-sites/marketplace/components/reviews-list/index.tsx]: Let's hide this behind a feature flag so that we can ship the editing functionality right away.

**Review by cpapazoglou — APPROVED** [review]
LGTM, except for `Notify me when my review is approved and published.` which we can hide behind a feature flag.

File: client/data/marketplace/use-marketplace-reviews.ts
Comment (WBerredo) [inline client/data/marketplace/use-marketplace-reviews.ts]: We may want to use the error callback to do this and the client may decide how to display the error.

File: client/data/marketplace/use-marketplace-reviews.ts
Comment (WBerredo) [inline client/data/marketplace/use-marketplace-reviews.ts]: We need to research a better way to invalidate other queries related to this product, as the average rating is not being updated after we edit a review.

**Review by WBerredo — COMMENTED** [review]
Trying to update I got this:

CleanShot 2024-01-02 at 11 28 01@2x

More backend related, but I don't think we need to verify if the user has a current subscription when they are updating a review, once we already do this when the review is created. For example: A user got a plugin and they made a review, later on they didn't like something and uninstall, but now they can't change their rating to reflect it.

---
I also haven't found any design about how to display errors, but I assume something similar to the "pending approval" warning (with different colors) should work:

image

File: client/data/marketplace/use-marketplace-reviews.ts
Comment (gcsecsey) [reply client/data/marketplace/use-marketplace-reviews.ts]: Added this here: https://github.com/Automattic/dotcom-forge/issues/4897

**Comment by gcsecsey** [comment]
> I also haven't found any design about how to display errors, but I assume something similar to the "pending approval" warning (with different colors) should work:

I'll wait for https://github.com/Automattic/dotcom-forge/issues/4824 to be merged with this as discussed.

File: client/data/marketplace/use-marketplace-reviews.ts
Comment (gcsecsey) [reply client/data/marketplace/use-marketplace-reviews.ts]: I removed the `onError` call, and added it here: https://github.com/Automattic/wp-calypso/pull/85305/files#diff-72b59db4ba804fcbd1962c7f5515d8e995f414cf742c6dd513d2ef1bbc15dad2R189

This works as expected, but I couldn't resolve a type mismatch:
CleanShot 2024-01-03 at 16 18 44@2x
CleanShot 2024-01-03 at 16 20 37@2x

@WBerredo could you help with how we could add the error type here? Thanks!

Comment (WBerredo) [reply client/data/marketplace/use-marketplace-reviews.ts]: Passing the correct type for `useMutation` should do the trick.


On `use-marketplace-reviews.ts` line `295`:
```typescript
return useMutation< MarketplaceReviewResponse, ErrorResponse, MarketplaceReviewBody >( {
```

Comment (gcsecsey) [reply client/data/marketplace/use-marketplace-reviews.ts]: Thanks @WBerredo, that really did the trick in https://github.com/Automattic/wp-calypso/pull/85305/commits/558dc914aed5ae680ac05027019841acb906a8a1, CI is green now.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/10685580

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @gcsecsey for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/10685580).

**Pull request merged**
